<a href="https://colab.research.google.com/github/SoIllEconomist/ds4b/blob/master/python_ds4b/01_exploration/02_data_transformation/02_data_transformation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Transformation

## Introduction

Visualisation is an important tool for insight generation, but it is rare that you get the data in exactly the right form you need. Often you’ll need to create some new variables or summaries, or maybe you just want to rename the variables or reorder the observations in order to make the data a little easier to work with. You’ll learn how to do all that (and more!) in this chapter, which will teach you how to transform your data using the dplyr package and a new dataset on flights departing New York City in 2013.
### Prerequisites
In this chapter we’re going to focus on how to use the dplyr package, another core member of the tidyverse. We’ll illustrate the key ideas using data from the NYC Flight data, and use `seaborn` to help us understand the data.

In [0]:
import pandas as pd
import numpy as np

flights = pd.read_csv("flights.csv")

### NYC Flights Dataset

To explore the basic data manipulation with `pandas`. The U.S. Department of Transportation's (DOT) Bureau of Transportation Statistics tracks the on-time performance of domestic flights operated by large air carriers. Summary information on the number of on-time, delayed, canceled, and diverted flights is published in DOT's monthly Air Travel Consumer Report and in this dataset of 2015 flight delays and cancellations. The data comes from the [US Bureau of Transportation Statistics](https://www.kaggle.com/usdot/flight-delays#flights.csv).

You might notice that this data frame prints a differently from other data frames you might have used in the past: it only shows the first few rows and all the columns that fit on one screen. (To see the whole dataset, you can run `flights` which will open the dataset instead of `flights.head()`. 

In [4]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150000 entries, 0 to 149999
Data columns (total 24 columns):
Unnamed: 0        150000 non-null int64
year              150000 non-null int64
month             150000 non-null int64
day               150000 non-null int64
dep_time          146274 non-null float64
sched_dep_time    150000 non-null int64
dep_delay         146274 non-null float64
arr_time          146138 non-null float64
sched_arr_time    150000 non-null int64
arr_delay         145890 non-null float64
carrier           150000 non-null object
flight            150000 non-null int64
tail_num          148864 non-null object
origin            150000 non-null object
dest              150000 non-null object
air_time          145890 non-null float64
distance          150000 non-null int64
hour              150000 non-null int64
minute            150000 non-null int64
time_hour         150000 non-null object
gain              145890 non-null float64
speed             146138 non-nul

You might have noticed that `.info()` prints a concise summary of a DataFrame.

This method prints information about a DataFrame including the index dtype and column dtypes, non-null values and memory usage. 

## Pandas Basics

In this chapter you are going to learn the key pandas methods and funcgtions that allow you to solve the vast majority of your data manipulation challenges:

1. Pick observations by their values.
1. Reorder the rows.
1. Pick variables by their names.
1. Create new variables with functions of existing variables.
1. Collapse many values down to a single summary.

These can all be used in conjunction with `groupby()` which changes the scope of each function from operating on the entire dataset to operating on it group-by-group. These six functions provide the verbs for a language of data manipulation.

### Query

`query()` allows you to subset observations based on their values. The first argument is the name of the data frame. The second and subsequent arguments are the expressions that filter the data frame. For example, we can select all flights on January 1st with:

In [5]:
flights.query("month ==1 & day == 1")

Unnamed: 0  year  month  day  ...  gain      speed     hours  gain_per_hour
72             291  2013      1    1  ...   1.0  44.112792  2.783333       0.359281
255            666  2013      1    1  ...   4.0  68.115672  5.700000       0.701754
742            185  2013      1    1  ...   7.0  46.965174  2.416667       2.896552
845             67  2013      1    1  ...   8.0  66.819604  2.733333       2.926829
1807           108  2013      1    1  ...  15.0  12.425249  0.633333      23.684211
...            ...   ...    ...  ...  ...   ...        ...       ...            ...
149308         687  2013      1    1  ...   1.0  21.364486  2.250000       0.444444
149424         299  2013      1    1  ...   4.0  11.267176  0.833333       4.800000
149521          31  2013      1    1  ...   8.0  71.478261  2.550000       3.137255
149635         198  2013      1    1  ...   2.0  52.190784  2.683333       0.745342
149733         107  2013      1    1  ...  15.0  13.333333  0.633333      23.684211

[711 rows x 24 columns]

When you run that line of code, pandas executes the querying operation and returns a new data frame. pandas functions never modify their inputs, so if you want to save the result, you’ll need to use the assignment operator, `=`:

In [0]:
jan1 = flights.query("month ==1 & day == 1")

### Comparisons

To use querying effectively, you have to know how to select the observations that you want using the comparison operators. Python provides the standard suite: `>`, `>=`, `<`, `<=`, `!=` (not equal), and `==` (equal).

When you’re starting out with Python, the easiest mistake to make is to use = instead of == when testing for equality. When this happens you’ll get an informative error:

In [7]:
flights.query("month =1")

ValueError: ignored

There’s another common problem you might encounter when using ==: floating point numbers. These results might surprise you!

In [0]:
from math import sqrt

In [9]:
sqrt(2) ** 2 == 2

False

In [10]:
1/49 * 49 == 1

False

Computers use finite precision arithmetic (they obviously can’t store an infinite number of digits!) so remember that every number you see is an approximation.

## Logical Operators

Multiple arguments to `query()` are combined with “and”: every expression must be true in order for a row to be included in the output. For other types of combinations, you’ll need to use Boolean operators yourself: `&` is “and”, `|` is “or”, and `~` is “not”. The figure below shows the complete set of Boolean operations.

![Complete set of boolean operations. x is the left-hand circle, y is the right-hand circle, and the shaded region show which parts each operator selects.](https://github.com/SoIllEconomist/ds4b/blob/master/python_ds4b/01_exploration/02_data_transformation/transform_logical.png?raw=1)



In [11]:
flights.query("month in [11, 12]")

Unnamed: 0  year  month  ...        speed     hours  gain_per_hour
1            69716  2013     11  ...    38.359739  3.400000       2.352941
5            83149  2013     11  ...  1143.809524  5.516667       3.806647
9            59523  2013     11  ...    83.239437  2.783333       5.389222
19           91707  2013     12  ...    30.826211  1.833333     -15.818182
23           61079  2013     11  ...     9.897611  0.933333      16.071429
...            ...   ...    ...  ...          ...       ...            ...
149989       93287  2013     12  ...    37.710407  3.633333       1.100917
149993      104629  2013     12  ...    31.294343  1.883333      -6.902655
149994       70680  2013     11  ...    44.448858  1.983333       1.008403
149995       91738  2013     12  ...    52.315036  2.783333     -14.371257
149998       98756  2013     12  ...   771.578947  1.733333      15.000000

[46793 rows x 24 columns]

Sometimes you can simplify complicated subsetting by remembering De Morgan’s law: `!(x & y)` is the same as `!x | !y`, and `!(x | y)` is the same as `!x & !y`. For example, if you wanted to find flights that weren’t delayed (on arrival or departure) by more than two hours, you could use either of the following two filters:

In [12]:
flights.query("~(arr_delay > 120 | dep_delay > 120)")

Unnamed: 0  year  month  ...        speed     hours  gain_per_hour
0           155090  2013      3  ...   131.270125  5.566667       2.514970
1            69716  2013     11  ...    38.359739  3.400000       2.352941
2           159787  2013      3  ...    45.411955  2.383333       0.839161
3           122785  2013      2  ...    27.282609  1.550000       3.870968
4           167006  2013      4  ...  1210.188679  2.550000       0.392157
...            ...   ...    ...  ...          ...       ...            ...
149994       70680  2013     11  ...    44.448858  1.983333       1.008403
149995       91738  2013     12  ...    52.315036  2.783333     -14.371257
149996      141274  2013      3  ...    15.785582  1.183333      10.140845
149997      141586  2013      3  ...   399.500000  3.383333      -2.955665
149999       43654  2013     10  ...    76.250617  5.866667       0.000000

[146175 rows x 24 columns]

In [13]:
flights.query("arr_delay <= 120 | dep_delay <= 120") # Double Check

Unnamed: 0  year  month  ...        speed     hours  gain_per_hour
0           155090  2013      3  ...   131.270125  5.566667       2.514970
1            69716  2013     11  ...    38.359739  3.400000       2.352941
2           159787  2013      3  ...    45.411955  2.383333       0.839161
3           122785  2013      2  ...    27.282609  1.550000       3.870968
4           167006  2013      4  ...  1210.188679  2.550000       0.392157
...            ...   ...    ...  ...          ...       ...            ...
149994       70680  2013     11  ...    44.448858  1.983333       1.008403
149995       91738  2013     12  ...    52.315036  2.783333     -14.371257
149996      141274  2013      3  ...    15.785582  1.183333      10.140845
149997      141586  2013      3  ...   399.500000  3.383333      -2.955665
149999       43654  2013     10  ...    76.250617  5.866667       0.000000

[143391 rows x 24 columns]

Whenever you start using complicated, multipart expressions in query(), consider making them explicit variables instead. That makes it much easier to check your work. You’ll learn how to create new variables shortly.

### Missing Values

One important feature of R that can make comparison tricky are missing values, or `NA`s (“not availables”). `NA` represents an unknown value so missing values are “contagious”: almost any operation involving an unknown value will also be unknown or False.

In [14]:
np.nan

nan

In [15]:
np.nan > 5

False

In [16]:
np.nan < 5

False

In [17]:
np.nan + 10

nan

In [18]:
np.nan / 2

nan

The most confusing result is this one:

In [19]:
np.nan == np.nan

False

It’s easiest to understand why this is true with a bit more context:

In [20]:
# Let x be Mary's age. We don't know how old she is.
x = np.nan

# Let y be John's age. We don't know how old he is.
y = np.nan

# Are John and Mary the same age?
x == y
# We don't know!

False

If you want to determine if a value is missing, use `pd.isna()`

In [21]:
pd.isna(x)

True

`query()` only includes rows where the condition is TRUE; it excludes both FALSE.

### Exercises

1. Find all the flights that
  1. Had an arrival delay of two or more hours
  1. Flew to Houston (IAH or HOU)
  1. Were operated by United, American, or Delta
  1. Departed in summer (July, August, and September)
  1. Arrived more than two hours late, but didn’t leave late
  1. Were delayed by at least an hour, but made up over 30 minutes in flight
  1. Departed between midnight and 6am (inclusive)6am

1. How many flights have a missing dep_time? What other variables are missing? What might these rows represent?

## Arrange rows with `sort_values()`
`sort_values()` works similarly to `query()` except that instead of selecting rows, it changes their order. It takes a data frame and a set of column names (or more complicated expressions) to order by. If you provide more than one column name, each additional column will be used to break ties in the values of preceding columns:


In [22]:
flights.sort_values(by=["year", "month", "day"])

Unnamed: 0  year  month  day  ...  gain       speed     hours  gain_per_hour
72             291  2013      1    1  ...   1.0   44.112792  2.783333       0.359281
255            666  2013      1    1  ...   4.0   68.115672  5.700000       0.701754
742            185  2013      1    1  ...   7.0   46.965174  2.416667       2.896552
845             67  2013      1    1  ...   8.0   66.819604  2.733333       2.926829
1807           108  2013      1    1  ...  15.0   12.425249  0.633333      23.684211
...            ...   ...    ...  ...  ...   ...         ...       ...            ...
149126      110889  2013     12   31  ... -23.0   30.951724  2.400000      -9.583333
149143      111237  2013     12   31  ...   2.0   27.791649  2.766667       0.722892
149203      111187  2013     12   31  ...  10.0   26.332553  2.433333       4.109589
149395      111073  2013     12   31  ...  13.0   43.727225  3.350000       3.880597
149429      110521  2013     12   31  ...  14.0  213.541203  3.200000       4.375000

[150000 rows x 24 columns]

Use `ascending=False` to re-order by a column in descending order:

In [23]:
flights.sort_values(by=["year", "month", "day"], ascending=False).head()

Unnamed: 0  year  month  day  ...  gain      speed     hours  gain_per_hour
123       111192  2013     12   31  ...  -5.0  69.082814  6.133333      -0.815217
348       111108  2013     12   31  ...  14.0  31.003495  2.566667       5.454545
356       110963  2013     12   31  ... -29.0  47.284768  4.150000      -6.987952
392       111040  2013     12   31  ...  18.0   7.285974  0.566667      31.764706
1099      110753  2013     12   31  ...  -2.0  21.211042  1.200000      -1.666667

[5 rows x 24 columns]

Missing values are sorted at the end:

In [24]:
flights.sort_values(by=["year", "month", "dep_time"]).tail()

Unnamed: 0  year  month  day  ...  gain  speed  hours  gain_per_hour
149437       96010  2013     12   14  ...   NaN    NaN    NaN            NaN
149609       87965  2013     12    5  ...   NaN    NaN    NaN            NaN
149650       87904  2013     12    5  ...   NaN    NaN    NaN            NaN
149791       98776  2013     12   17  ...   NaN    NaN    NaN            NaN
149961       88966  2013     12    6  ...   NaN    NaN    NaN            NaN

[5 rows x 24 columns]

Puttingi NAs first:

In [25]:
flights.sort_values(by=["year", "month", "dep_time"], na_position='first')

Unnamed: 0  year  month  day  ...  gain       speed     hours  gain_per_hour
367          24250  2013      1   28  ...   NaN         NaN       NaN            NaN
442          13964  2013      1   16  ...   NaN         NaN       NaN            NaN
601          13980  2013      1   16  ...   NaN         NaN       NaN            NaN
630          22535  2013      1   26  ...   NaN         NaN       NaN            NaN
1940         24226  2013      1   28  ...   NaN         NaN       NaN            NaN
...            ...   ...    ...  ...  ...   ...         ...       ...            ...
145417      109531  2013     12   29  ... -25.0  188.366534  3.333333      -7.500000
30948        95379  2013     12   13  ...   9.0  221.944444  3.200000       2.812500
82972       100795  2013     12   19  ...   7.0  220.921659  3.216667       2.176166
120577       91492  2013     12    9  ...   9.0  221.944444  3.250000       2.769231
148448       87893  2013     12    5  ...  14.0  224.543326  3.033333       4.615385

[150000 rows x 24 columns]

### Exercises
1. How could you use `sort_values()` to sort all missing values to the start?.

1. Sort flights to find the most delayed flights. Find the flights that left earliest.

1. Sort flights to find the fastest flights.

1. Which flights travelled the longest? Which travelled the shortest?



## Select columns with `[]`, `loc`, and `filter()`

It’s not uncommon to get datasets with hundreds or even thousands of variables. In this case, the first challenge is often narrowing in on the variables you’re actually interested in. Bracket notation `[]` allows you to rapidly zoom in on a useful subset using operations based on the names of the variables.

Bracket notation is not terribly useful with the flights data because we only have 19 variables, but you can still get the general idea:

In [26]:
flights[['year','month','day']]

year  month  day
0       2013      3   21
1       2013     11   15
2       2013      3   26
3       2013      2   14
4       2013      4    2
...      ...    ...  ...
149995  2013     12   10
149996  2013      3    6
149997  2013      3    6
149998  2013     12   17
149999  2013     10   18

[150000 rows x 3 columns]

In [27]:
flights.filter(["year", "month", "day"])

year  month  day
0       2013      3   21
1       2013     11   15
2       2013      3   26
3       2013      2   14
4       2013      4    2
...      ...    ...  ...
149995  2013     12   10
149996  2013      3    6
149997  2013      3    6
149998  2013     12   17
149999  2013     10   18

[150000 rows x 3 columns]

In [28]:
flights.loc[:,'year':'day']

year  month  day
0       2013      3   21
1       2013     11   15
2       2013      3   26
3       2013      2   14
4       2013      4    2
...      ...    ...  ...
149995  2013     12   10
149996  2013      3    6
149997  2013      3    6
149998  2013     12   17
149999  2013     10   18

[150000 rows x 3 columns]

There are a number of helper functions you can use:

```python
df.filter(regex="^abc") # starts with abc
df.filter(regex="$xyz") # ends with xyz
df.filter(regex="example") # contains example
```

## rename

In [29]:
df = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]})
df

A  B
0  1  4
1  2  5
2  3  6

Rename columns using a mapping:

In [30]:
df.rename(columns={"A": "a", "B": "c"})

a  c
0  1  4
1  2  5
2  3  6

Rename index using a mapping:

In [31]:
df.rename(index={0: "x", 1: "y", 2: "z"})

A  B
x  1  4
y  2  5
z  3  6

Using axis-style parameters

In [32]:
df.rename(str.lower, axis='columns')

a  b
0  1  4
1  2  5
2  3  6

In [33]:
df.rename({1: 2, 2: 4}, axis='index')

A  B
0  1  4
2  2  5
4  3  6

If you want any changes to stick, you need to use the `inplace=True` arguments.

In [0]:
df.rename(str.lower, axis="columns", inplace=True)

In [35]:
df # Now we see that everything is changed.

a  b
0  1  4
1  2  5
2  3  6

In [0]:
flights.rename({"tailnum":"tail_num"}, axis="columns", inplace=True)

In [37]:
flights.columns

Index(['Unnamed: 0', 'year', 'month', 'day', 'dep_time', 'sched_dep_time',
       'dep_delay', 'arr_time', 'sched_arr_time', 'arr_delay', 'carrier',
       'flight', 'tail_num', 'origin', 'dest', 'air_time', 'distance', 'hour',
       'minute', 'time_hour', 'gain', 'speed', 'hours', 'gain_per_hour'],
      dtype='object')

### Exercises
1. Brainstorm as many ways as possible to select dep_time, dep_delay, arr_time, and arr_delay from flights.

1. What happens if you include the name of a variable multiple times in a selection statement call?

1. What does the filter and loc methods do? Why might it be helpful in conjunction with this list?
```python
vars = ["year", "month", "day", "dep_delay", "arr_delay"]
```

## Add new variables

Besides selecting sets of existing columns, it’s often useful to add new columns that are functions of existing columns.

In [0]:
flights["gain"] = flights.dep_delay - flights.arr_delay
flights["speed"] = flights.distance / flights.arr_time * 60
flights['hours'] = flights.air_time / 60
flights["gain_per_hour"] = flights.gain / flights.hours

### Exercises
1. Currently dep_time and `sched_dep_time` are convenient to look at, but hard to compute with because they’re not really continuous numbers. Convert them to a more convenient representation of number of minutes since midnight.

1. Compare air_time with `arr_time` - `dep_time`. What do you expect to see? What do you see? What do you need to do to fix it?

1. Compare `dep_time`, `sched_dep_time`, and `dep_delay`. How would you expect those three numbers to be related?

1. Find the 10 most delayed flights using a ranking function. How do you want to handle ties?

## Grouped summaries: split-apply-combine

In this section, you'll learn how to use the pandas groupby operation, which draws from the well-known split-apply-combine strategy.

Intuitively, you want to split the dataset into groups, one for each year, and then to compute a summary statistic, such as the mean or the median, and then to see whether this statistic increases over the years (after this, you may want to perform a statistical test).

The framework is known as **split-apply-combine** because we:

1. **split** the data into groups by creating a groupby object from the original DataFrame;
1. **apply** a function, in this case, an aggregation function that computes a summary statistic (you can also transform or filter your data in this step);
1. **combine** the results into a new Dataframe

This is the conceptual framework for the analysis at hand. We'll also necessarily delve into groupby objects, which are not the most intuitive objects. The process of **split-apply-combine** with groupby objects is a pattern that we all perform intuitively, as we'll see, but it took Hadley Wickham to formalize the procedure in 2011 with his paper The Split-Apply-Combine Strategy for Data Analysis.

Here you'll use pandas, groupby objects and the principles of split-apply-combine to check out the flights dataset.

### Summarising your data with plots and statistics

The pandas DataFrame .info() method is invaluable. Applying it below shows that you have 150000 rows and 23 columns of data, but also that the column of interest, `dep_delay`, has only 146274 non-null values. This means that there are 3726 missing values:

In [40]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150000 entries, 0 to 149999
Data columns (total 24 columns):
Unnamed: 0        150000 non-null int64
year              150000 non-null int64
month             150000 non-null int64
day               150000 non-null int64
dep_time          146274 non-null float64
sched_dep_time    150000 non-null int64
dep_delay         146274 non-null float64
arr_time          146138 non-null float64
sched_arr_time    150000 non-null int64
arr_delay         145890 non-null float64
carrier           150000 non-null object
flight            150000 non-null int64
tail_num          148864 non-null object
origin            150000 non-null object
dest              150000 non-null object
air_time          145890 non-null float64
distance          150000 non-null int64
hour              150000 non-null int64
minute            150000 non-null int64
time_hour         150000 non-null object
gain              145890 non-null float64
speed             146138 non-nul

If you'd like to check out several summary statistics of the DataFrame, you can also do this using the `.describe()` method:

In [41]:
flights.describe()

Unnamed: 0      year  ...          hours  gain_per_hour
count  150000.000000  150000.0  ...  145890.000000  145890.000000
mean    88738.232053    2013.0  ...       2.563177       2.821006
std     51217.980586       0.0  ...       1.575603      10.060148
min         0.000000    2013.0  ...       0.333333    -169.230769
25%     44364.750000    2013.0  ...       1.416667      -1.682243
50%     88874.500000    2013.0  ...       2.233333       2.666667
75%    133126.250000    2013.0  ...       3.250000       7.328244
max    177350.000000    2013.0  ...      11.583333      85.384615

[8 rows x 19 columns]

### Groupbys and split-apply-combine to answer the question

#### Step 1. Split

Now that you've checked out out data, it's time for the fun part. You'll first use a `.groupby()` method to split the data into groups, where each group is the date. This is the split in split-apply-combine:

In [0]:
flights_by_month = flights[["year", "month", "gain_per_hour", "dep_delay"]].groupby(by=["year", "month"])

This creates a *groupby* object:

In [44]:
type(flights_by_day)

pandas.core.groupby.generic.DataFrameGroupBy

#### Step 2. Apply
Such groupby objects are very useful. Remember that the `.describe()` method for a DataFrame returns summary statistics for numeric columns? Well, the `.describe()` method for DataFrameGroupBy objects returns summary statistics for each numeric column, but computed for each group in the split. In your case, it's for each release_year. This is an example of the apply in split-apply-combine: you're applying the `.describe()` method to each group in the groupby. Do this and print the first 5 rows of the result:

In [67]:
flights_by_month.describe().head()

gain_per_hour                       ... dep_delay              
                   count      mean        std  ...       50%   75%     max
year month                                     ...                        
2013 1           22354.0  2.407132  10.188676  ...      -2.0   8.0  1301.0
     2           19987.0  2.768578   9.998214  ...      -2.0   9.0   853.0
     3           23626.0  3.991077  10.779158  ...      -1.0  12.0   911.0
     4           10186.0  2.282663  10.473078  ...      -2.0  10.0   639.0
     10          24156.0  3.776325   9.650829  ...      -3.0   4.0   390.0

[5 rows x 16 columns]

If you want to see what the grouping looks like, you can pass the groupby object to the function list():

In [63]:
list(flights_by_month)[5] # Cast grouping as a list and check out one month

((2013, 11),
         Unnamed: 0  year  month  ...        speed     hours  gain_per_hour
 1            69716  2013     11  ...    38.359739  3.400000       2.352941
 5            83149  2013     11  ...  1143.809524  5.516667       3.806647
 9            59523  2013     11  ...    83.239437  2.783333       5.389222
 23           61079  2013     11  ...     9.897611  0.933333      16.071429
 25           61471  2013     11  ...    61.093248  2.333333       5.142857
 ...            ...   ...    ...  ...          ...       ...            ...
 149980       76392  2013     11  ...  1435.714286  2.400000       3.333333
 149981       72730  2013     11  ...    53.955409  2.500000       5.200000
 149982       58576  2013     11  ...    61.192982  2.383333       0.419580
 149983       69658  2013     11  ...    29.194729  2.383333       2.097902
 149994       70680  2013     11  ...    44.448858  1.983333       1.008403
 
 [22974 rows x 24 columns])

### Step 3. Combine
Let's say that you wanted the mean or median *dep_delay* and *gain_per_hour* for each month. Then you can apply the `.mean()` or `.median()` method, respectively, to the groupby object and 'combine' these into a new DataFrame.

In [0]:
flights_med_by_month = flights_by_month.median()

In [74]:
flights_med_by_month.head()

gain_per_hour  dep_delay
year month                          
2013 1           2.201835       -2.0
     2           2.654867       -2.0
     3           3.797468       -1.0
     4           2.337662       -2.0
     10          3.290177       -3.0

In [75]:
flights_med_by_month.unstack()

gain_per_hour                                ... dep_delay               
month            1         2         3         4   ...        4    10   11   12
year                                               ...                         
2013       2.201835  2.654867  3.797468  2.337662  ...      -2.0 -3.0 -3.0  0.0

[1 rows x 14 columns]

### Groupbys and split-apply-comnine in Daily Use

Groupby objects are not intuitive. They do, however, correspond to a natural the act of splitting a dataset with respect to one its columns (or more than one, but let's save that for another post about grouping by multiple columns and hierarchical indexes).

The split-apply-combine principle is not only elegant and practical, it's something that Data Scientists use daily, as in the above example.

## Exercises
1. Brainstorm at least 5 different ways to assess the typical delay characteristics of a group of flights. Consider the following scenarios:

  1. A flight is 15 minutes early 50% of the time, and 15 minutes late 50% of the time.

  1. A flight is always 10 minutes late.

  1. A flight is 30 minutes early 50% of the time, and 30 minutes late 50% of the time.

  1. 99% of the time a flight is on time. 1% of the time it’s 2 hours late.

Which is more important: arrival delay or departure delay?

1. Our definition of cancelled flights `(flights.dep_delay.isna() | flights.arr_delay.isna())` is slightly suboptimal. Why? Which is the most important column?

1. Look at the number of cancelled flights per day. Is there a pattern? Is the proportion of cancelled flights related to the average delay?

1. Which carrier has the worst delays? Challenge: can you disentangle the effects of bad airports vs. bad carriers? Why/why not? (Hint: think about `flights.groupby(['carrier','dest']).count()`

1. What does the sort argument to count() do. When might you use it?

1. Which plane (tailnum) has the worst on-time record?

1. What time of day should you fly if you want to avoid delays as much as possible?

1. For each destination, compute the total minutes of delay. For each flight, compute the proportion of the total delay for its destination.

1. Look at each destination. Can you find flights that are suspiciously fast? (i.e. flights that represent a potential data entry error). Compute the air time a flight relative to the shortest flight to that destination. Which flights were most delayed in the air?

1. Find all destinations that are flown by at least two carriers. Use that information to rank the carriers.

1. For each plane, count the number of flights before the first delay of greater than 1 hour.